File: 278_firstBadVersion2.py
'''
    The isBadVersion API is already defined for you.
    def isBadVersion(version: int) -> bool:
    
    candidate qn: what if there is no bad version?

'''

class Solution:
    def helper(self, lo: int, hi: int) -> int:
        
        mid = int((lo + hi) / 2)
        # print(f'lo: {lo}  mid: {mid}  hi: {hi}')
        if isBadVersion(mid):
            # if the mid is bad, then first bad is in [lo, mid - 1]
            # if mid is bad but not first bad then we move forward
            if isBadVersion(mid - 1):
                return self.helper(1, mid - 1, n)
            else:
                return mid
        else:
            # if the mid is not bad, then first bad is in [mid + 1, hi]
            return self.helper(mid + 1, hi)

    def firstBadVersion(self, n: int) -> int:
        return self.helper(1, n)


def isBadVersion(n: int) -> bool:
    return (n >= 69)

if __name__ == '__main__':

    soln = Solution()
    print(f'first bad version is {soln.firstBadVersion(100)}')

File: 278_firstBadVersion2.java
public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        return helper(1, n);
    }

    private int helper(int lo, int hi) {
        int mid = (lo + hi) / 2;
        if (isBadVersion(mid)) {
            // if the mid is bad, then first bad is in [lo, mid - 1]
            // if mid is bad but not first bad then we move forward
            if (isBadVersion(mid - 1)) {
                return helper(lo, mid - 1);
            } else {
                return mid;
            }
        } else {
            // if the mid is not bad, then bad is in [mid + 1, hi]
            return helper(mid + 1, hi);
        }
    }
}